topic com.hazelcast.core.HazelcastException: Failed to publish message: wrapping com.hazelcast.core.OperationTimeoutException

using version 3.8

running multi tests in parallel, lots of thread, lots of structures, some async puts, get's,  lots of load,  so the timeOutException is expected

`com.hazelcast.core.HazelcastException`  
is wrapping 
`java.util.concurrent.ExecutionException` 
which is wrapping 
`com.hazelcast.core.OperationTimeoutException`

```
at com.hazelcast.topic.impl.reliable.ReliableTopicProxy.publish(ReliableTopicProxy.java:172)
    at hzcmd.topic.Publisher.timeStep(Publisher.java:13)
    at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
    at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
    at remote.bench.BenchThread.call(BenchThread.java:38)
    at remote.bench.BenchThread.call(BenchThread.java:12)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.ExecutionException: AddOperation got rejected before execution due to not starting within the operation-call-timeout of: 60000 ms. Current time: 2017-02-22 08:56:33.975. Start time: 2017-02-22 08:54:54.936. Total elapsed time: 99039 ms. Invocation{op=com.hazelcast.ringbuffer.impl.operations.AddOperation{serviceName='hz:impl:ringbufferService', identityHash=2084149414, partitionId=54, replicaIndex=0, callId=-9838688, invocationTime=1487746494936 (2017-02-22 08:54:54.936), waitTimeout=-1, callTimeout=60000, name=_hz_rb_topicA_topic}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1487746494936, firstInvocationTime='2017-02-22 08:54:54.936', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[10.212.1.117]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:151)
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:99)
    at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:75)
    at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
    at com.hazelcast.topic.impl.reliable.ReliableTopicProxy.addWithBackoff(ReliableTopicProxy.java:190)
    at com.hazelcast.topic.impl.reliable.ReliableTopicProxy.publish(ReliableTopicProxy.java:162)
    ... 11 more
Caused by: com.hazelcast.core.OperationTimeoutException: AddOperation got rejected before execution due to not starting within the operation-call-timeout of: 60000 ms. Current time: 2017-02-22 08:56:33.975. Start time: 2017-02-22 08:54:54.936. Total elapsed time: 99039 ms. Invocation{op=com.hazelcast.ringbuffer.impl.operations.AddOperation{serviceName='hz:impl:ringbufferService', identityHash=2084149414, partitionId=54, replicaIndex=0, callId=-9838688, invocationTime=1487746494936 (2017-02-22 08:54:54.936), waitTimeout=-1, callTimeout=60000, name=_hz_rb_topicA_topic}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1487746494936, firstInvocationTime='2017-02-22 08:54:54.936', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[10.212.1.117]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
    ... 17 more
```

@vbekiaris 
I see it’s deep in the invocation system; I ‘ll take a look a bit later to see if we can peel the `ExecutionException` and throw the original cause or we need to wrap it as-is

--- a/hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientReliableTopicProxy.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientReliableTopicProxy.java
@@ -20,7 +20,6 @@ import com.hazelcast.client.config.ClientReliableTopicConfig;
 import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.spi.ClientProxy;
 import com.hazelcast.core.ExecutionCallback;
-import com.hazelcast.core.HazelcastException;
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
 import com.hazelcast.core.ICompletableFuture;
 import com.hazelcast.core.ITopic;
@@ -51,6 +50,7 @@ import java.util.concurrent.Executor;
 
 import static com.hazelcast.ringbuffer.impl.RingbufferService.TOPIC_RB_PREFIX;
 import static com.hazelcast.topic.impl.reliable.ReliableTopicService.SERVICE_NAME;
+import static com.hazelcast.util.ExceptionUtil.peel;
 import static com.hazelcast.util.Preconditions.checkNotNull;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
@@ -114,10 +114,9 @@ public class ClientReliableTopicProxy<E> extends ClientProxy implements ITopic<E
                 default:
                     throw new IllegalArgumentException("Unknown overloadPolicy:" + overloadPolicy);
             }
-        } catch (RuntimeException e) {
-            throw e;
         } catch (Exception e) {
-            throw new HazelcastException("Failed to publish message: " + payload + " to topic:" + name, e);
+            throw (RuntimeException) peel(e, null,
+                    "Failed to publish message: " + payload + " to topic:" + getName());
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/ReliableTopicProxy.java
+++ b/hazelcast/src/main/java/com/hazelcast/topic/impl/reliable/ReliableTopicProxy.java
@@ -43,6 +43,7 @@ import java.util.concurrent.Executor;
 
 import static com.hazelcast.ringbuffer.impl.RingbufferService.TOPIC_RB_PREFIX;
 import static com.hazelcast.spi.ExecutionService.ASYNC_EXECUTOR;
+import static com.hazelcast.util.ExceptionUtil.peel;
 import static com.hazelcast.util.Preconditions.checkNotNull;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
@@ -166,10 +167,9 @@ public class ReliableTopicProxy<E> extends AbstractDistributedObject<ReliableTop
             }
 
             localTopicStats.incrementPublishes();
-        } catch (RuntimeException e) {
-            throw e;
         } catch (Exception e) {
-            throw new HazelcastException("Failed to publish message: " + payload + " to topic:" + getName(), e);
+            throw (RuntimeException) peel(e, null,
+                    "Failed to publish message: " + payload + " to topic:" + getName());
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/util/ExceptionUtil.java
+++ b/hazelcast/src/main/java/com/hazelcast/util/ExceptionUtil.java
@@ -53,6 +53,24 @@ public final class ExceptionUtil {
     }
 
     public static <T extends Throwable> Throwable peel(final Throwable t, Class<T> allowedType) {
+        return peel(t, allowedType, null);
+    }
+
+    /**
+     * Processes {@code Throwable t} so that the returned {@code Throwable}'s type matches {@code allowedType} or
+     * {@code RuntimeException}. Processing may include unwrapping {@code t}'s cause hierarchy, wrapping it in a
+     * {@code HazelcastException} or just returning the same instance {@code t} if it is already an instance of
+     * {@code RuntimeException}.
+     *
+     * @param t             {@code Throwable} to be peeled
+     * @param allowedType   the type expected to be returned; when {@code null}, this method returns instances
+     *                      of {@code RuntimeException}
+     * @param message       if not {@code null}, used as the message in the {@code HazelcastException} that
+     *                      may wrap the peeled {@code Throwable}
+     * @param <T>           expected type of {@code Throwable}
+     * @return              the peeled {@code Throwable}
+     */
+    public static <T extends Throwable> Throwable peel(final Throwable t, Class<T> allowedType, String message) {
         if (t instanceof RuntimeException) {
             return t;
         }
@@ -70,7 +88,11 @@ public final class ExceptionUtil {
             return t;
         }
 
-        return new HazelcastException(t);
+        if (message != null) {
+            return new HazelcastException(message, t);
+        } else {
+            return new HazelcastException(t);
+        }
     }
 
     public static RuntimeException rethrow(final Throwable t) {

